backends/bmv2/common/extern.h

```diff
@@ -0,0 +1,154 @@
+            registerExternConverter(#extern_name, this); }                      \
+        static ExternConverter_##extern_name##__VA_ARGS__ singleton;            \
+        void convertExternInstance(ConversionContext* ctxt,             \
+            const IR::Declaration* c, const IR::ExternBlock* eb) override; };
+
+#define EXTERN_CONVERTER_W_INSTANCE(extern_name, ...)                          \
+    class ExternConverter_##extern_name##__VA_ARGS__ : public ExternConverter { \
+        ExternConverter_##extern_name##__VA_ARGS__() {                          \
+            registerExternConverter(#extern_name, this); }                      \
+        static ExternConverter_##extern_name##__VA_ARGS__ singleton;            \
+        void convertExternInstance(ConversionContext* ctxt,             \
+            const IR::Declaration* c, const IR::ExternBlock* eb) override; };
+
+#define EXTERN_CONVERTER_W_OBJECT_AND_INSTANCE_AND_MODEL(extern_name, type, name, ...)  \
+    class ExternConverter_##extern_name##__VA_ARGS__ : public ExternConverter { \
+        type&      name;                                                        \
+        ExternConverter_##extern_name##__VA_ARGS__() :                          \
+            name(type::instance) {                                              \
+            registerExternConverter(#extern_name, this); }                      \
+        static ExternConverter_##extern_name##__VA_ARGS__ singleton;            \
+        void convertExternInstance(ConversionContext* ctxt,             \
+            const IR::Declaration* c, const IR::ExternBlock* eb) override;      \
+        Util::IJson* convertExternObject(ConversionContext* ctxt,               \
+            const P4::ExternMethod* em, const IR::MethodCallExpression* mc,     \
+            const IR::StatOrDecl* s) override; };
+
+#define EXTERN_CONVERTER_W_OBJECT_AND_INSTANCE(extern_name, ...)               \
+    class ExternConverter_##extern_name##__VA_ARGS__ : public ExternConverter { \
+        ExternConverter_##extern_name##__VA_ARGS__() {                          \
+            registerExternConverter(#extern_name, this); }                      \
+        static ExternConverter_##extern_name##__VA_ARGS__ singleton;            \
+        void convertExternInstance(ConversionContext* ctxt,             \
+            const IR::Declaration* c, const IR::ExternBlock* eb) override;      \
+        Util::IJson* convertExternObject(ConversionContext* ctxt,               \
+            const P4::ExternMethod* em, const IR::MethodCallExpression* mc,     \
+            const IR::StatOrDecl* s) override; };
+
+#define EXTERN_CONVERTER_SINGLETON(extern_name, ...)                           \
+    ExternConverter_##extern_name##__VA_ARGS__                                 \
+       ExternConverter_##extern_name##__VA_ARGS__::singleton;
+
+#define CONVERT_EXTERN_INSTANCE(extern_name, ...)                                     \
```

Comment:
I don't like macros much. In particular, these macros hardwire some parameter names. When you read the use of the macro you have no idea what these things are or where they come from (e.g., eb, em, mc).
If you use macros maybe you can at least expose all the parameters in the macro, and perhaps give longer names to the parameters.

Reply:
I will think of a better way to write this macro, maybe in the next PR.